File: src/main/java/com/pboucher/mailSender/service/impl/MailServiceImpl.java
package com.pboucher.mailSender.service.impl;

import com.pboucher.mailSender.metier.Contact;
import com.pboucher.mailSender.service.MailService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.mail.SimpleMailMessage;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
public class MailServiceImpl implements MailService {

    private static final String EMAIL_SECRET_SANTA = "secretsanta@pierreboucher.fr";

    @Autowired
    private JavaMailSender emailSender;

    @Override
    public void sendSimpleMail(HashMap<Contact, Contact> contacts, int montant) {
        for (Map.Entry<Contact, Contact> contact: contacts.entrySet()) {
            SimpleMailMessage message = emailTemplate(contact.getKey(), contact.getValue(), montant);
            emailSender.send(message);
        }
    }

    @Override
    public HashMap<Contact, Contact> shuffleContacts(List<Contact> contacts) {
        HashMap<Contact, Contact> res = new HashMap<>();
        List<Contact> randomContacts = new ArrayList<>(contacts);

        randomContacts = shuffleList(randomContacts);

        for (int i = 0; i<contacts.size(); i++) {
            res.put(contacts.get(i), randomContacts.get(i));
        }

        return res;
    }

    /**
     * Collections.shuffle pouvais laisser un contact envoyer un cadeau a lui meme
     *
     * @param contacts
     * @return
     */
    private List<Contact> shuffleList(List<Contact> contacts) {
        List<Contact> oldContact = new ArrayList<>(contacts);
        Collections.shuffle(contacts);
        List<Contact> save = new ArrayList<>(contacts);

        for (int i = 0; i < contacts.size(); i++) {
            if (oldContact.get(i) == contacts.get(i)) {
                if (i < contacts.size()-1) {
                    contacts.set(i, save.get(i+1));
                    contacts.set(i+1, save.get(i));
                } else if (i == contacts.size()-1) {
                    contacts.set(i, save.get(0));
                    contacts.set(0, save.get(i));
                }
            }
        }

        return contacts;
    }

    private SimpleMailMessage emailTemplate(Contact firstContact, Contact secondContact, int montant) {
        SimpleMailMessage message = new SimpleMailMessage();
        message.setTo(firstContact.getEmail());
        message.setFrom(EMAIL_SECRET_SANTA);
        message.setSubject("Secret Santa | Père noel secret");
        message.setText("Chère " + firstContact.getNom() +"\n Tu devra offrir un cadeau à " + secondContact.getNom() +" d'un montant maximum de " + montant + "€");
        return message;
    }
}

File: src/main/kotlin/com/pboucher/mailsender/service/impl/MailServiceImpl.kt
package com.pboucher.mailsender.service.impl

import com.pboucher.mailsender.metier.Contact
import com.pboucher.mailsender.service.MailService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.mail.SimpleMailMessage
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.mail.javamail.JavaMailSenderImpl
import org.springframework.stereotype.Service
import java.util.*
import kotlin.collections.HashMap


@Service
class MailServiceImpl: MailService {

    private val EMAIL_SECRET_SANTA = "secretsanta@pierreboucher.fr"

    @Autowired
    private lateinit var emailSender: JavaMailSender

    override fun sendSimpleMail(contacts: HashMap<Contact, Contact>, montant: Int) {
        for ((key, value) in contacts) {
            val message: SimpleMailMessage = emailTemplate(key, value, montant)
            emailSender.send(message)
        }
    }

    override fun shuffleContacts(contacts: List<Contact>): HashMap<Contact, Contact> {
        val res = HashMap<Contact, Contact>()
        var randomContacts: MutableList<Contact> = ArrayList(contacts)
        randomContacts = shuffleList(randomContacts)
        for (i in contacts.indices) {
            res[contacts[i]] = randomContacts[i]
        }
        return res
    }

    /**
     * Collections.shuffle pouvais laisser un contact envoyer un cadeau a lui meme
     *
     * @param contacts
     * @return
     */
    private fun shuffleList(contacts: MutableList<Contact>): MutableList<Contact> {
        val oldContact: List<Contact> = ArrayList(contacts)
        Collections.shuffle(contacts)
        val save: List<Contact> = ArrayList(contacts)
        for (i in contacts.indices) {
            if (oldContact[i] == contacts[i]) {
                if (i < contacts.size - 1) {
                    contacts[i] = save[i + 1]
                    contacts[i + 1] = save[i]
                } else if (i == contacts.size - 1) {
                    contacts[i] = save[0]
                    contacts[0] = save[i]
                }
            }
        }
        return contacts
    }

    private fun emailTemplate(firstContact: Contact, secondContact: Contact, montant: Int): SimpleMailMessage {
        val message = SimpleMailMessage()
        message.setTo(firstContact.email)
        message.setFrom(EMAIL_SECRET_SANTA)
        message.setSubject("Secret Santa | Père noel secret")
        message.setText(
            "Chère ${firstContact.nom}, Tu devra offrir un cadeau à ${secondContact.nom} d'un montant maximum de ${montant}€".trimMargin()
        )
        return message
    }
}
